### PR TITLE
[SPARK-8582][Core]Optimize checkpointing to avoid computing an RDD twice

### PR DESCRIPTION
**Pull request by zsxwing**
Unlike #7021, this PR uses an approach similar to `persist` to compute and checkpoint an RDD at the same job. When computing an RDD at the first time, each partition will be checkpointed and read back as  an iterator (maybe compute it again, not sure which one is better). After finishing the job, we also check if all partitions are checkpointed, if not, we still need to launch a job to checkpoint the missing partitions.

### CODE REVIEW

**Comment by zsxwing** [comment]
I feel confused for `synchronized` in `RDDCheckpointData`: since RDD is not thread safe, why we need to make `RDDCheckpointData` thread safe?

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -258,6 +258,16 @@ abstract class RDD[T: ClassTag](
    * subclasses of RDD.
    */
   final def iterator(split: Partition, context: TaskContext): Iterator[T] = {
+    if (!isCheckpointedAndMaterialized) {
```
Comment (ryan-williams) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: just out of curiosity, any reason not to do an `if`/`else` here?
```
if (!isCheckpointedAndMaterialized && 
    checkpointData.exists(_.isInstanceOf[ReliableRDDCheckpointData[T]])) {
  SparkEnv.get.checkpointMananger.getOrCompute(
    this, checkpointData.get.asInstanceOf[ReliableRDDCheckpointData[T]], split, context)
} else {
  computeOrReadCache(split, context)
}
```

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 2]
```diff
@@ -81,6 +81,8 @@ class SparkEnv (
   @deprecated("Actor system is no longer supported as of 1.4.0", "1.4.0")
   val actorSystem: ActorSystem = _actorSystem
 
+  private[spark] val checkpointMananger = new CheckpointManager
```
Comment (ryan-williams) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala]: typo: `s/Mananger/Manager/g`, jfyi

File: core/src/main/scala/org/apache/spark/CheckpointManager.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+import org.apache.spark.rdd.{ReliableRDDCheckpointData, ReliableCheckpointRDD, RDD}
+import org.apache.spark.storage._
+
+class CheckpointManager extends Logging {
+
+  /** Keys of RDD partitions that are being checkpointed. */
+  private val checkpointingRDDPartitions = new mutable.HashSet[RDDBlockId]
+
+  /** Gets or computes an RDD partition. Used by RDD.iterator() when an RDD is cached. */
```
Comment (ryan-williams) [inline core/src/main/scala/org/apache/spark/CheckpointManager.scala]: does "cached" still make sense here?

**Comment by tdas** [comment]
@andrewor14 Can you please take a look at this?

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
There is still one problem in this PR. Currently, if an RDD is not persisted, this RDD will be recomputed to generate an Iterator after checkpointing (We cannot use the original Iterator since it's consumed). However, recomputing RDD may be slower than reading from the checkpoint file if the RDD is very complicated. Since we don't know whether recomputing is faster than reading from the checkpoint file, maybe we should add an option to the `checkpoint` API and let the user make the decision.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
@zsxwing I took a quick look and I have a high level question. Why not just do the checkpointing iterator? IIUC this approach involves reading the iterator back from disk to return the values. Wouldn't that be potentially expensive? Also, this doesn't fix it for local checkpointing.

If we have a general checkpointing iterator, then RDD doesn't have to change much and we don't need to introduce another `CheckpointManager`, which I find a little clunky.

**Comment by zsxwing** [comment]
@andrewor14 I just opened #9428 to take over #7021 and fix potential issues in the previous PR.

**Pull request closed without merging**